### PR TITLE
CHK-9 addition of checkexperiments script

### PR DESCRIPTION
**Pull request by gabdank**
We are adding a script "checkexperiments.py" to the checkfiles repo.
This script will run after checkfiles is done. The goal of this script is to switch started experiments into submited status, if raw data is there. Current version will only report to the bot-channel.

### CODE REVIEW

File: checkexperiments.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
+"""\
```
Comment (caseylitton) [inline checkexperiments.py]: get rid of the extra \

File: checkexperiments.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+Run status check on experiments.
+
+Example.
+
+    %(prog)s --username ACCESS_KEY_ID --password SECRET_ACCESS_KEY \\
+        --out experiments_statuses.log https://www.encodeproject.org
+"""
+import datetime
+import json
```
Comment (caseylitton) [inline checkexperiments.py]: remove import json since it's not used.

File: checkexperiments.py [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+import copy
+import os.path
+import subprocess
+from urllib.parse import urljoin
+import requests
+from slackclient import SlackClient
+
+EPILOG = __doc__
+
+def run(out, err, url, username, password, search_query, accessions_list=None, bot_token=None, dry_run=False):
```
Comment (caseylitton) [inline checkexperiments.py]: Need two empty lines above the function.

File: checkexperiments.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+                        as_user=True)
+
+        sc.api_call(
+            "chat.postMessage",
+            channel="#bot-reporting",
+            text=finishing_run,
+            as_user=True
+        )
+
+def main():
```
Comment (caseylitton) [inline checkexperiments.py]: Need two empty lines above function declarations.

File: checkexperiments.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+                            file_date = datetime.datetime.strptime(
+                                file_obj['date_created'][:10], "%Y-%m-%d")
+                            dates.append(file_date)
+                            files.append(file_obj)
+                            if file_obj.get('read_count') and file_obj.get('replicate'):
+                                if not replicates_reads.get(file_obj.get('replicate')) is None:
+                                    replicates_reads[file_obj.get('replicate')] += \
+                                        file_obj.get('read_count')
+        except requests.exceptions.RequestException as e:
+            print (e)
```
Comment (caseylitton) [inline checkexperiments.py]: remove extra space between 'print' and it's argument '(e)'

File: checkexperiments.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+                                    experiment_accession + '\t' + ex['status'] + \
+                                    '\t-> submitted\t' + max(dates).strftime("%Y-%m-%d") + '\n')
+                                out.flush()
+                            else:
+                                err.write(
+                                    award_obj.get('rfa') + '\t' +
+                                    experiment_accession + '\taudit errors\n')
+                                err.flush()
+
+
```
Comment (caseylitton) [inline checkexperiments.py]: Too many blank lines.  Only one blank line a row in functions.  

In general, I would not leave any blank lines in a function without really good reason.  Indenting in python takes care of visually separating code blocks so we don't need to add extra spaces.

File: checkexperiments.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+                    if award_obj.get('rfa') == 'modENCODE':
+                        for rep in replicates_reads:
+                            if replicates_reads[rep] < min_depth['modENCODE-chip']:
+                                depth_flag = True
+                                err.write(
+                                    award_obj.get('rfa') + '\t' + \
+                                    experiment_accession + '\t' + rep + \
+                                    '\treads_count=' + str(replicates_reads[rep]) + \
+                                    '\texpected count=' + \
+                                    str(min_depth['modENCODE-chip']) + '\n')
```
Comment (caseylitton) [inline checkexperiments.py]: Use string format.
```
                                msg = (
                                    '{}\t{}\t{}'
                                    '\ttreads_count={}'
                                    '\texpected_count={}\n'.format(
                                        award_obj.get('rfa'),
                                        experiment_accession,
                                        rep,
                                        str(replicates_reads[rep]),
                                        str(min_depth['modENCODE-chip'])
                                    )
                                )
                                err.write(msg)
```
Much easier to read and update.

File: checkexperiments.py [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+                        if ex['assay_term_name'] in min_depth:
+                            for rep in replicates_reads:
+                                if replicates_reads[rep] < min_depth[ex['assay_term_name']]:
+                                    depth_flag = True
+                                    err.write(
+                                        award_obj.get('rfa') + '\t' + \
+                                        experiment_accession + '\t' + rep + \
+                                        '\treads_count=' + \
+                                        str(replicates_reads[rep]) + '\texpected count=' + \
+                                        str(min_depth[ex['assay_term_name']]) + '\n')
```
Comment (caseylitton) [inline checkexperiments.py]: Again use string format.

File: checkexperiments.py [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+                        replicates.add(replicate_obj['@id'])
+                        replicates_reads[replicate_obj['@id']] = 0
+                if  ex.get('files'):
+                    for file_acc in ex.get('files'):
+                        file_request = session.get(urljoin(
+                            url,
+                            file_acc + '?frame=object&format=json'))
+                        file_obj = file_request.json()
+                        if file_obj.get('file_format') == 'fastq' and \
+                           file_obj.get('status') not in ['uploading',
```
Comment (caseylitton) [inline checkexperiments.py]: put the list in the condition in a variable outside the for loop.
```
allowed_status = ['uploading', 'content error', 'upload failed']
```
Also multiline conditionals should be inside parenthesis and the second condition should be indented.
```
                        if (file_obj.get('file_format') == 'fastq' and
                                file_obj.get('status') not in allowed_status):
```

File: checkexperiments.py [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
+            if ex.get('replicates'):
+                for replicate in ex.get('replicates'):
+                    replicate_request = session.get(urljoin(
+                        url,
+                        replicate + '?frame=object&format=json'))
+                    replicate_obj = replicate_request.json()
+                    if replicate_obj.get('status') not in ['deleted']:
+                        replicates.add(replicate_obj['@id'])
+                        replicates_reads[replicate_obj['@id']] = 0
+                if  ex.get('files'):
```
Comment (caseylitton) [inline checkexperiments.py]: extra space in the condition, after the if.

File: checkexperiments.py [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+                return
+            else:
+                local = copy.deepcopy(r.json()['@graph'])
+                graph.extend(local)
+    # checkexperiments using a query
+    else:
+        r = session.get(
+            urljoin(
+                url,
+                '/search/?type=Experiment' \
```
Comment (caseylitton) [inline checkexperiments.py]: The \ is not needed inside brackets

File: checkexperiments.py [hunk 12]
```diff
@@ -0,0 +1,266 @@
...
+    version = '0.11'
+
+    try:
+        ip_output = subprocess.check_output(
+            ['hostname'], stderr=subprocess.STDOUT).strip()
+        ip = ip_output.decode(errors='replace').rstrip('\n')
+    except subprocess.CalledProcessError as e:
+        ip = ''
+
+    initiating_run = 'STARTING Checkexperiments version ' + \
```
Comment (caseylitton) [inline checkexperiments.py]: We should mix format and string concatenation.   The whole thing should just be put in a string with a format function.
```
    initiating_run = (
        'STARTING Checkexperiments version '
        '{} ({}) ({}): {} on {} at {}'.format(
            version, url, search_query, dr, ip, datetime.datetime.now()
        )
    )
```

**Review by caseylitton — CHANGES_REQUESTED** [review]
There were a bunch pep8 issues.  

I suggest just always using the format function for any string concatenation that involves variables.  It's much easier to read, write, and update once you get in the habit of using it.

File: checkexperiments.py (hunk 1 above)
Comment (gabdank) [reply checkexperiments.py]: done

File: checkexperiments.py (hunk 2 above)
Comment (gabdank) [reply checkexperiments.py]: done

File: checkexperiments.py (hunk 3 above)
Comment (gabdank) [reply checkexperiments.py]: done

File: checkexperiments.py (hunk 12 above)
Comment (gabdank) [reply checkexperiments.py]: done

File: checkexperiments.py (hunk 11 above)
Comment (gabdank) [reply checkexperiments.py]: done

File: checkexperiments.py (hunk 10 above)
Comment (gabdank) [reply checkexperiments.py]: done

File: checkexperiments.py (hunk 9 above)
Comment (gabdank) [reply checkexperiments.py]: done (these are not "allowed_status" but rather "excluded_statuses")

File: checkexperiments.py (hunk 5 above)
Comment (gabdank) [reply checkexperiments.py]: done

File: checkexperiments.py (hunk 7 above)
Comment (gabdank) [reply checkexperiments.py]: done

File: checkexperiments.py (hunk 6 above)
Comment (gabdank) [reply checkexperiments.py]: done

File: checkexperiments.py (hunk 4 above)
Comment (gabdank) [reply checkexperiments.py]: done

**Comment by gabdank** [comment]
@caseylitton I have modified the code according to your suggestions

**Review by caseylitton — APPROVED** [review]

**Pull request merged**